src/web/rustdoc.rs

```diff
@@ -812,4 +842,277 @@ mod test {
             Ok(())
         })
     }
+
+    #[test]
+    fn platform_links_go_to_current_path() {
+        fn get_platform_links(
+            path: &str,
+            web: &TestFrontend,
+        ) -> Result<Vec<(String, String)>, failure::Error> {
+            use html5ever::tendril::TendrilSink;
+            assert_success(path, web)?;
+            let data = web.get(path).send()?.text()?;
+            let dom = kuchiki::parse_html().one(data);
+            Ok(dom
+                .select(r#"a[aria-label="Platform"] + ul li a"#)
+                .expect("invalid selector")
+                .map(|el| {
+                    let url = el
+                        .attributes
+                        .borrow()
+                        .get("href")
+                        .expect("href")
+                        .to_string();
+                    let name = el.text_contents();
+                    (name, url)
+                })
+                .collect())
+        }
+
+        wrapper(|env| {
+            let (db, web) = (env.db(), env.frontend());
+
+            // no explicit default-target
+            db.fake_release()
+                .name("dummy")
+                .version("0.1.0")
+                .rustdoc_file("dummy/index.html", b"some content")
+                .rustdoc_file("dummy/struct.Dummy.html", b"some other content")
+                .add_target("x86_64-unknown-linux-gnu")
+                .create()?;
+
+            assert_eq!(
+                get_platform_links("/dummy/0.1.0/dummy/", web)?,
+                vec![(
+                    "x86_64-unknown-linux-gnu".into(),
+                    "/dummy/0.1.0/dummy/index.html".into()
+                ),],
+            );
+
+            assert_eq!(
+                get_platform_links("/dummy/0.1.0/dummy/index.html", web)?,
+                vec![(
+                    "x86_64-unknown-linux-gnu".into(),
+                    "/dummy/0.1.0/dummy/index.html".into()
+                ),],
+            );
```

Comment:
Perhaps it would be helpful to write a test abstraction that does some of this boilerplate for you? Something like this:

```
fn test_platforms(default_target: Option<&str>, platforms: &[&str], pages: &[&str]) {
    let mut builder = db.fake_release().name("dummy").version("0.1.0");
    for page in pages {
        builder = builder.rustdoc_file(page, b"dummy content");
    }
    for platform in platforms {
        for page in pages {
            builder = builder.rustdoc_file(format!("{}/{}", platform, page), b"dummy content");
        }
        builder = builder.add_target(platform);
    }
    if let Some(default) = default_target {
        builder = builder.default_target(default);
    }
    builder.create()
}
```
That would shrink down the boilerplate, making it easier to read and write tests.

Reply:
The builder boilerplate doesn't seem too bad to me, it's the assertions that are really annoying. I can probably simplify them a _little_ with a custom assertion macro (one day `track(caller)` will be stable and we can start writing nice assertion functions).